datafusion/common/src/scalar.rs

```diff
@@ -50,9 +50,52 @@ use arrow::{
 use arrow_array::cast::as_list_array;
 use arrow_array::{ArrowNativeTypeOp, Scalar};
 
-/// Represents a dynamically typed, nullable single value.
-/// This is the single-valued counter-part to arrow's [`Array`].
+/// A dynamically typed, nullable single value, (the single-valued counter-part
```

Comment:
I think the rust purists would say at compile time the compiler knows it is `ScalarValue` (rather than say `int32`)

So in order to do anything you have to "dynamically dispatch", aka 
```rust
// dispatch on type
match scalar {
  ScalarValue::Int32(...) => ...
  ScalarValue::Int64(...) => ...

...
}
```